server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -539,50 +551,69 @@ static ReplicaItemExecutionMode replicaItemExecutionMode(final BulkItemRequest r
         final SourceToParse sourceToParse =
             SourceToParse.source(request.index(), request.type(), request.id(), request.source(), request.getContentType())
                 .routing(request.routing());
-        try {
-            // if a mapping update is required to index this request, issue a mapping update on the master, and abort the
-            // current indexing operation so that it can be retried with the updated mapping from the master
-            // The early abort uses the RetryOnPrimaryException, but any other exception would be fine as well.
-            return primary.applyIndexOperationOnPrimary(request.version(), request.versionType(), sourceToParse,
-                request.getAutoGeneratedTimestamp(), request.isRetry(), update -> {
-                    mappingUpdater.updateMappings(update, primary.shardId(), sourceToParse.type());
-                    throw new ReplicationOperation.RetryOnPrimaryException(primary.shardId(), "Mapping updated");
-                });
-        } catch (ReplicationOperation.RetryOnPrimaryException e) {
-            return primary.applyIndexOperationOnPrimary(request.version(), request.versionType(), sourceToParse,
-                request.getAutoGeneratedTimestamp(), request.isRetry(), update -> mappingUpdater.verifyMappings(update, primary.shardId()));
+        Engine.IndexResult result;
+
+        result = primary.applyIndexOperationOnPrimary(request.version(), request.versionType(), sourceToParse,
```

Comment:
this looks pretty much like the code in the delete part below. Can we maybe break it out in a shared routine  and pass a closure to it to actually process the operation.

Reply:
+1. I'll do this once the other discussion settles.